mixins/labelled-mixin.js

```diff
@@ -0,0 +1,87 @@
+
+export const LabelMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			_label: { type: String, reflect: true }
+		};
+	}
+
+};
+
+export const LabelledMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			labelledBy: { type: String, reflect: true, attribute: 'labelled-by' },
+			label: { type: String }
+		};
+	}
+
+	updated(changedProperties) {
+		super.updated(changedProperties);
+
+		if (!changedProperties.has('labelledBy')) return;
+
+		if (this._labelObserver) this._labelObserver.disconnect();
+
+		if (this.labelledBy) {
+
+			let labelElem = this.getRootNode().querySelector(`#${this.labelledBy}`);
+			const ancestor = this._getCommonAncestor(labelElem);
+
+			const getLabel = () => {
+				if (labelElem.tagName.includes('-')) return labelElem._label;
+				else return labelElem.textContent;
+			};
+
+			this._labelObserver = new MutationObserver((mutations) => {
+				for (let i = 0; i < mutations.length; i++) {
+
+					for (let j = 0; j < mutations[i].removedNodes.length; j++) {
+						if (mutations[i].removedNodes[j] === labelElem) {
+							labelElem = null;
+							break;
+						}
+					}
+
+					if (!labelElem && mutations[i].addedNodes.length > 0) {
+						labelElem = this.getRootNode().querySelector(`#${this.labelledBy}`);
+						break;
+					}
+
+				}
```

Comment:
Not sure if this works, but `forEach` and an `indexOf` might simplify the looping logic a bit.

```suggestion
				mutations.forEach((m) => {

					if (m.removedNodes.indexOf(labelElem)) {
						labelElem = null;
						return;
					}

					if (!labelElem && m.addedNodes.length > 0) {
						labelElem = this.getRootNode().querySelector(`#${this.labelledBy}`);
						break;
					}

				});
```

Reply:
I'll include this in my changes... need a couple modifications since `NodeList` does not support `indexOf`.  Would not want the `return` statement, and the `break` statement should become `return`.